src/ServiceControl/Operations/ErrorIngestion.cs

```diff
@@ -20,9 +20,9 @@ public ErrorIngestion(ErrorIngestor errorIngestor, string errorQueue, RawEndpoin
             importFailuresHandler = new SatelliteImportFailuresHandler(documentStore, loggingSettings, onCriticalError);
         }
 
-        public async Task EnsureStarted()
+        public async Task EnsureStarted(CancellationToken cancellationToken)
         {
-            await startStopSemaphore.WaitAsync().ConfigureAwait(false);
+            await startStopSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Would it make sense to include this into the try / catch finally to make sure we always release?

Reply:
I wondered about this. What's the semantics of releasing a semaphore we don't hold? If this method throws, do we have the lock?

Reply:
In all the other code paths where we have a semaphore we always have the WaitAsync and Release pattern in a try finally because that is how it is intended to be used.

There is only one case when the counter is not decrement

https://referencesource.microsoft.com/#mscorlib/system/threading/SemaphoreSlim.cs,612

This happens when you enter with the token already canceled.

Release has explicit lock semantics

https://referencesource.microsoft.com/#mscorlib/system/threading/SemaphoreSlim.cs,771

so it is guaranteed those operations are in sync. Therefore going with WaitAsync finally Release seems to be the most straight forward way